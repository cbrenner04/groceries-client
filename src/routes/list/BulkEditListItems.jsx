import React from 'react';
import PropTypes from 'prop-types';
import Async from 'react-async';

import { fetchItemsToEdit } from './utils';
import BulkEditListItemsForm from './containers/BulkEditListItemsForm';
import Loading from '../../components/Loading';
import UnknownError from '../error_pages/UnknownError';

function BulkEditListItems(props) {
  return (
    <Async
      promiseFn={fetchItemsToEdit}
      listId={props.match.params.list_id}
      itemType={props.match.params[0]}
      search={props.location.search}
      history={props.history}
    >
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data) => (
          <BulkEditListItemsForm
            history={props.history}
            list={data.list}
            lists={data.lists}
            items={data.items}
            categories={data.categories}
            listUsers={data.listUsers}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
}

BulkEditListItems.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      list_id: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  location: PropTypes.shape({
    search: PropTypes.string.isRequired,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
};

export default BulkEditListItems;
