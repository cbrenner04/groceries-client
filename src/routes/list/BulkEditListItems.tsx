import React from 'react';
import Async, { type PromiseFn } from 'react-async';
import { useLocation, useNavigate, useParams } from 'react-router';

import Loading from 'components/Loading';
import type { IList, IListItem, IListUser } from 'typings';

import { fetchItemsToEdit } from './utils';
import BulkEditListItemsForm from './containers/BulkEditListItemsForm';
import UnknownError from '../error_pages/UnknownError';

const BulkEditListItems: React.FC = (): React.JSX.Element => {
  const navigate = useNavigate();
  const { list_id: listId } = useParams();
  const location = useLocation();

  return (
    <Async
      promiseFn={fetchItemsToEdit as unknown as PromiseFn<void>}
      listId={listId}
      search={location.search}
      navigate={navigate}
    >
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data: {
          list: IList;
          lists: IList[];
          items: IListItem[];
          categories: string[];
          list_users: IListUser[];
        }): React.JSX.Element => (
          <BulkEditListItemsForm
            navigate={navigate}
            list={data.list}
            lists={data.lists}
            items={data.items}
            categories={data.categories}
            listUsers={data.list_users}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
};

export default BulkEditListItems;
