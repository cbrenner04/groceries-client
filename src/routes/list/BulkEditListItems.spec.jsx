import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import BulkEditListItems from './BulkEditListItems';
import axios from '../../utils/api';

describe('BulkEditListItems', () => {
  const props = {
    match: {
      params: {
        0: 'grocery_list_items',
        list_id: 'id1',
      },
    },
    location: {
      search: '?item_ids=id1,id2',
    },
  };
  const renderBulkEditListItems = (newProps) => {
    const history = createMemoryHistory();
    return render(
      <Router history={history}>
        <BulkEditListItems {...newProps} history={history} />
      </Router>,
    );
  };

  it('renders the Loading component when fetch request is pending', () => {
    const { container, getByText } = renderBulkEditListItems(props);
    const status = getByText('Loading...');

    expect(container).toMatchSnapshot();
    expect(status).toBeTruthy();
  });

  it('displays UnknownError when an error occurs', async () => {
    axios.get = jest.fn().mockRejectedValue({ message: 'failed to send request' });
    const { container, getByRole } = renderBulkEditListItems(props);
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(container).toMatchSnapshot();
    expect(getByRole('button')).toHaveTextContent('refresh the page');
  });

  it('displays BulkEditListItems', async () => {
    axios.get = jest.fn().mockResolvedValue({
      data: {
        items: [
          {
            user_id: 'id1',
            id: 'id1',
            product: 'foo',
          },
        ],
        list: {
          id: 'id1',
          type: 'GroceryList',
          name: 'foo',
          created_at: 'some date',
          completed: false,
          refreshed: false,
          owner_id: 'id1',
        },
        list_users: [{ id: 'id1', email: 'foo@example.com' }],
        categories: [],
        lists: [
          {
            id: 'id1',
            type: 'GroceryList',
            name: 'foobar',
            created_at: 'some date',
            completed: false,
            refreshed: false,
            owner_id: 'id1',
          },
        ],
      },
    });
    const { container, getByText } = renderBulkEditListItems(props);
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(container).toMatchSnapshot();
    expect(getByText('Update Items')).toBeVisible();
  });
});
