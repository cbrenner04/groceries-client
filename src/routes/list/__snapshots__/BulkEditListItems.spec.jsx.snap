// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkEditListItems displays BulkEditListItems 1`] = `
<div>
  <h1>
    Edit 
    foo
  </h1>
  <br />
  <form
    autocomplete="off"
    class=""
  >
    <div>
      Move or copy these items to another list.
    </div>
    <br />
    <div
      class="form-check-inline"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="move-action-copy"
          name="copy"
          type="radio"
          value="false"
        />
        <label
          class="form-check-label"
          for="move-action-copy"
          title=""
        >
          Copy
        </label>
      </div>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="move-action-move"
          name="move"
          type="radio"
          value="false"
        />
        <label
          class="form-check-label"
          for="move-action-move"
          title=""
        >
          Move
        </label>
      </div>
    </div>
    <hr />
    <div>
      Update attributes for all items.
    </div>
    <br />
    <div>
      <label
        class="form-label"
        for="quantity"
      >
        Quantity
      </label>
      <input
        class="form-control"
        id="quantity"
        name="quantity"
        placeholder="3 bags"
        type="text"
        value=""
      />
      <div
        class="ms-1 mt-1"
      >
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="clearQuantity"
            name="clearQuantity"
            type="checkbox"
          />
          <label
            class="form-check-label"
            for="clearQuantity"
            title=""
          >
            Clear quantity
          </label>
        </div>
      </div>
    </div>
    <div>
      <label
        class="form-label"
        for="category"
      >
        Category
      </label>
      <input
        class="form-control"
        id="category"
        list="categories"
        name="category"
        type="text"
        value=""
      />
      <datalist
        id="categories"
      />
      <div
        class="ms-1 mt-1"
      >
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="clearCategory"
            name="clearCategory"
            type="checkbox"
          />
          <label
            class="form-check-label"
            for="clearCategory"
            title=""
          >
            Clear category
          </label>
        </div>
      </div>
    </div>
    <div
      class="d-grid gap-2 mt-3"
    >
      <button
        class="btn btn-success"
        type="submit"
      >
        Update Items
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`BulkEditListItems displays UnknownError when an error occurs 1`] = `
<div>
  <h1>
    Something went wrong!
  </h1>
  <h2>
    We are currently unable to render this page.
  </h2>
  <p>
    Please check your connection and try to Â 
    <button
      style="background-color: transparent; cursor: pointer; display: inline; margin: 0px; padding: 0px; color: rgb(0, 123, 255); text-decoration: none;"
    >
      refresh the page
    </button>
  </p>
</div>
`;

exports[`BulkEditListItems renders the Loading component when fetch request is pending 1`] = `
<div>
  <div
    style="display: flex; justify-content: center; align-items: center;"
  >
    <div
      class="spinner-border"
      role="status"
      style="width: 100px; height: 100px; position: absolute; top: 50%;"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
